src/hotspot/share/services/nmtPreInit.hpp

```diff
@@ -0,0 +1,353 @@
+  }
+
+  static NMTPreInitAllocation* find_in_map(void* p) {
+    assert(_table != NULL, "stray allocation?");
+    return _table->find(p);
+  }
+
+  static NMTPreInitAllocation* find_and_remove_in_map(void* p) {
+    assert(!_nmt_was_initialized, "lookup map cannot be modified after NMT initialization");
+    assert(_table != NULL, "stray allocation?");
+    return _table->find_and_remove(p);
+  }
+
+  // Just a wrapper for os::malloc to avoid including os.hpp here.
+  static void* do_os_malloc(size_t size);
+
+public:
+
+  // Switches from NMT pre-init state to NMT post-init state;
+  //  in post-init, no modifications to the lookup table are possible.
+  static void pre_to_post();
+
+  // Called from os::malloc.
+  // Returns true if allocation was handled here; in that case,
+  // *rc contains the return address.
+  static bool handle_malloc(void** rc, size_t size) {
+    size = MAX2((size_t)1, size);         // malloc(0)
+    if (!_nmt_was_initialized) {
+      // pre-NMT-init:
+      // Allocate entry and add address to lookup table
+      NMTPreInitAllocation* a = NMTPreInitAllocation::do_alloc(size);
+      add_to_map(a);
+      (*rc) = a->payload();
+      _num_mallocs_pre ++;
+      return true;
+    }
+    return false;
+  }
+
+  // Called from os::realloc.
+  // Returns true if reallocation was handled here; in that case,
+  // *rc contains the return address.
+  static bool handle_realloc(void** rc, void* old_p, size_t new_size) {
+    if (old_p == NULL) {                  // realloc(NULL, n)
+      return handle_malloc(rc, new_size);
+    }
+    new_size = MAX2((size_t)1, new_size); // realloc(.., 0)
+    if (!_nmt_was_initialized) {
+      // pre-NMT-init:
+      // - the address must already be in the lookup table
+      // - find the old entry, remove from table, reallocate, add to table
+      NMTPreInitAllocation* a = find_and_remove_in_map(old_p);
+      a = NMTPreInitAllocation::do_reallocate(a, new_size);
+      add_to_map(a);
+      (*rc) = a->payload();
+      _num_reallocs_pre++;
+      return true;
+    } else {
+      // post-NMT-init:
+      // If the old block was allocated during pre-NMT-init, we must relocate it: the
+      //  new block must be allocated with "normal" os::malloc.
+      // We do this by:
+      // - look up (but not remove! lu table is read-only here.) the old entry
+      // - allocate new memory via os::malloc()
+      // - manually copy the old content over
+      // - return the new memory
+      // - The lu table is readonly so we keep the old address in the table. And we leave
+      //   the old block allocated too, to prevent the libc from returning the same address
+      //   and confusing us.
+      NMTPreInitAllocation* a = find_in_map(old_p);
+      if (a != NULL) { // this was originally a pre-init allocation
+        void* p_new = do_os_malloc(new_size);
+        ::memcpy(p_new, a->payload(), MIN2(a->size, new_size));
+        (*rc) = p_new;
+        _num_reallocs_pre_to_post++;
```

Comment:
post-NMT-init counter updates are racy.

Reply:
True, this is racy. It's just for diagnostics though - I rather remove them than make them atomic since we would pay for this with every malloc. Or, maybe atomic + debug only?

Reply:
Either one is fine.